src/modelhub/utils/inference.py

```diff
@@ -145,3 +166,161 @@ def build_file_paths_for_prediction(
+
+    # Build new mask
+    selection_mask = apply_atom_selection_mask(atom_array, template_selection_list)
+    logging.info(
+        "Selected %d atoms for token-level templating with %d syntaxes",
+        int(np.sum(selection_mask)),
+        len([s for s in template_selection_list if s]),
+    )
+
+    # OR with existing annotation if present
+    if "is_input_file_templated" in atom_array.get_annotation_categories():
+        existing = atom_array.get_annotation("is_input_file_templated").astype(bool)
+        selection_mask = existing | selection_mask
+    atom_array.set_annotation("is_input_file_templated", selection_mask)
+    return atom_array
+
+
+def apply_ground_truth_conformer_selection(
+    atom_array: AtomArray, ground_truth_conformer_selection: list[str] | str | None
+) -> AtomArray:
+    """Apply ground-truth conformer policy selection with union semantics.
+
+    Behavior:
+    - Creates `ground_truth_conformer_policy` if missing and initializes to IGNORE.
+    - For selected atoms, sets policy to at least ADD without downgrading any
+      existing policy (e.g., preserves REPLACE if present).
+
+    Args:
+        atom_array: AtomArray to annotate.
+        ground_truth_conformer_selection: Selection string(s). Single strings are converted to lists. If None/empty, no-op.
+
+    Returns:
+        The same AtomArray with `ground_truth_conformer_policy` updated.
+    """
+    # Convert to list if needed
+    ground_truth_conformer_selection_list = (
+        as_list(ground_truth_conformer_selection)
+        if ground_truth_conformer_selection
+        else []
+    )
+
+    if not ground_truth_conformer_selection_list:
+        if (
+            "ground_truth_conformer_policy"
+            not in atom_array.get_annotation_categories()
+        ):
+            atom_array.set_annotation(
+                "ground_truth_conformer_policy",
+                np.full(
+                    len(atom_array), GroundTruthConformerPolicy.IGNORE, dtype=np.int8
+                ),
+            )
+        return atom_array
+
+    # Ensure annotation exists
+    if "ground_truth_conformer_policy" not in atom_array.get_annotation_categories():
+        atom_array.set_annotation(
+            "ground_truth_conformer_policy",
+            np.full(len(atom_array), GroundTruthConformerPolicy.IGNORE, dtype=np.int8),
+        )
+
+    selection_mask = apply_atom_selection_mask(
+        atom_array, ground_truth_conformer_selection_list
+    )
+    logging.info(
+        "Selected %d atoms for ground-truth conformer policy with %d syntaxes",
+        int(np.sum(selection_mask)),
+        len([s for s in ground_truth_conformer_selection_list if s]),
+    )
+
+    existing = atom_array.get_annotation("ground_truth_conformer_policy")
+    existing[selection_mask] = GroundTruthConformerPolicy.ADD
+    atom_array.set_annotation("ground_truth_conformer_policy", existing)
+
+    return atom_array
+
+
+def apply_conformer_and_template_selections(
```

Comment:
Excellent that this is now a function outside of the inference engine - that'l make it much easier to work with